src/game_api/state.cpp

```diff
@@ -694,6 +695,76 @@ void init_state_update_hook()
     }
 }
 
+ExecutableMemory g_heap_clone_redirect;
+using OnHeapClone = void(uint64_t heap_to, uint64_t heap_container_from);
+OnHeapClone* g_heap_clone_trampoline{nullptr};
+void HeapClone(uint64_t heap_to, uint64_t heap_container_from)
+{
+    uint64_t location = memory_read<uint64_t>(State::get().location);
+    StateMemory* state_from = reinterpret_cast<StateMemory*>(memory_read<uint64_t>(heap_container_from + 0x88) + location);
+    StateMemory* state_to = reinterpret_cast<StateMemory*>(heap_to + location);
+    heap_clone_event(ON::PRE_CLONE_HEAP, state_from, state_to);
+    // g_heap_clone_trampoline(heap_container_to, heap_container_from, heap_to);
+    // heap_clone_event(ON::POST_CLONE_HEAP, state_from,state_to);
+}
+
+// Original function params: clone_heap(ThreadStorageContainer to, ThreadStorageContainer from)
+// HeapContainer has heap1 and heap2 variables, and some sort of timer, that just increases constantly, I guess to handle the rollback and multi-threaded stuff
+// The rest of what HeapContainer has is unknown for now
+// After writing to a chosen storage from the content of `from->heap1`, sets `to->heap2` to the newly copied thread storage

```

Comment:
this has to copy more than just state (like prng)
also if you only want callback then figuring out which thread is which from `ThreadStorageContainer` would be enough, then you could just use detour instead of all of this asm

btw. I specially wrote not long ago `patch_and_redirect` function, for exactly this kind of assembly hack, so you don't have to write all the boring jumps there and back and copy game code yourself xd

Reply:
yeah it also copies more stuff, but honestly I don't know how to handle that well in a callback, maybe make a struct for the whole storage that has prng, state, etc.

The issue with figuring out which thread is, is that I don't really know which thread is going to be the destination thread, at the start of the function you only know that `to->heap2` pointer is gonna get overwritten at the with a pointer to the destination thread.
For getting to which ThreadStorage will be written to, the game code seems to do some lock and unlock stuff, which may affect the other threads, so I decided to not try to figure out to which one will be written to with my own code

I'll look into the `patch_and_redirect` functions

Reply:
Also, I don't know if I can share some of the ghidra decompiled code to show it